services/src/main/java/io/grpc/services/HealthCheckingLoadBalancerFactory.java

```diff
@@ -0,0 +1,399 @@
+          @Override
+          public void onMessage(final HealthCheckResponse response) {
+            syncContext.execute(new Runnable() {
+                @Override
+                public void run() {
+                  handleResponse(response);
+                }
+              });
+          }
+
+          @Override
+          public void onClose(final Status status, Metadata trailers) {
+            syncContext.execute(new Runnable() {
+                @Override
+                public void run() {
+                  handleStreamClosed(status);
+                }
+              });
+          }
+        };
+    private final LoadBalancer delegate;
+    private final SynchronizationContext syncContext;
+    private final ScheduledExecutorService timerService;
+
+    private Subchannel subchannel;
+
+    // Set when the RPC is started. Cleared when the RPC is fully closed.
+    @Nullable
+    private ClientCall<HealthCheckRequest, HealthCheckResponse> activeCall;
+
+    // States about the most recently started RPC
+    private String lastCallServiceName;
+    private long lastCallStartNanos;
+    private boolean lastCallHasResponded;
+
+    // The service name should be used for health checking
+    private String serviceName;
+    private BackoffPolicy backoffPolicy;
+    // The state from the underlying Subchannel
+    private ConnectivityStateInfo rawState = ConnectivityStateInfo.forNonError(IDLE);
+    // The state concluded from health checking
+    private ConnectivityStateInfo concludedState = ConnectivityStateInfo.forNonError(IDLE);
+    // true if a health check stream should be kept.  Either there is an active RPC, or a retry is
+    // pending.
+    private boolean running;
+    // true if server returned UNIMPLEMENTED
+    private boolean disabled;
+    private ScheduledHandle retryTimer;
+
+    HealthCheckState(
+        LoadBalancer delegate, SynchronizationContext syncContext,
+        ScheduledExecutorService timerService) {
+      this.delegate = checkNotNull(delegate, "delegate");
+      this.syncContext = checkNotNull(syncContext, "syncContext");
+      this.timerService = checkNotNull(timerService, "timerService");
+    }
+
+    void handleResponse(HealthCheckResponse response) {
+      backoffPolicy = null;
+      lastCallHasResponded = true;
+      if (!running) {
+        return;
+      }
+      if (Objects.equal(response.getStatus(), ServingStatus.SERVING)) {
+        gotoState(ConnectivityStateInfo.forNonError(READY));
+      } else {
+        gotoState(
+            ConnectivityStateInfo.forTransientFailure(
+                Status.UNAVAILABLE.withDescription(
+                    "Health-check service responded "
+                    + response.getStatus() + " for '" + lastCallServiceName + "'")));
+      }
+      activeCall.request(1);
+    }
+
+    void handleStreamClosed(Status status) {
+      activeCall = null;
+      if (Objects.equal(status.getCode(), Code.UNIMPLEMENTED)) {
```

Comment:
Why is this before the `running` check? Seems like it could set `disabled = true` for the wrong TCP connection.

Reply:
I have moved the per-RPC states into a separate class. It's swapped out when the RPC has finished. Thus `handleStreamClosed()` is not called when `running == false`.